File: archive/2026-04-05/ch06_tree/t21_is_valid_bst.py
"""LeetCode 98. Validate Binary Search Tree
https://leetcode.cn/problems/validate-binary-search-tree/
"""
import sys
from typing import Optional

from tree_node import TreeNode


class Solution:
    pre_num: int = - sys.maxsize - 1

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        if not self.isValidBST(root.left):
            return False

        if root.val <= self.pre_num:
            return False
        self.pre_num = root.val

        return self.isValidBST(root.right)


if __name__ == "__main__":
    solution = Solution()
    root1 = TreeNode(2, TreeNode(1), TreeNode(3))
    assert solution.isValidBST(root1) is True

    solution = Solution()
    root2 = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    assert solution.isValidBST(root2) is False

    print("All tests passed.")

File: archive/2024-08-01/tree/T21_isValidBST.java
package org.rereview2024summer.tree;

public class T21_isValidBST {
    TreeNode preNode = null;
    public boolean isValidBST(TreeNode root) {
        if (root == null)
            return true;
        if (root.left != null)
            if (!isValidBST(root.left))
                return false;
        if (preNode != null && preNode.val >= root.val)
           return false;
        preNode = root;
        if (root.right != null)
            if (!isValidBST(root.right))
                return false;
        return true;
    }
}
